overlord/servicestate/service_control_test.go

```diff
@@ -0,0 +1,454 @@
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package servicestate_test
+
+import (
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"testing"
+	"time"
+
+	. "gopkg.in/check.v1"
+
+	"github.com/snapcore/snapd/dirs"
+	"github.com/snapcore/snapd/overlord"
+	"github.com/snapcore/snapd/overlord/servicestate"
+	"github.com/snapcore/snapd/overlord/snapstate"
+	"github.com/snapcore/snapd/overlord/state"
+	"github.com/snapcore/snapd/snap"
+	"github.com/snapcore/snapd/snap/snaptest"
+	"github.com/snapcore/snapd/systemd"
+	"github.com/snapcore/snapd/testutil"
+)
+
+func TestServiceControl(t *testing.T) { TestingT(t) }
+
+type serviceControlSuite struct {
+	testutil.BaseTest
+	state      *state.State
+	o          *overlord.Overlord
+	se         *overlord.StateEngine
+	serviceMgr *servicestate.ServiceManager
+	sysctlArgs [][]string
+}
+
+var _ = Suite(&serviceControlSuite{})
+
+const servicesSnapYaml1 = `name: test-snap
+version: 1.0
+apps:
+  someapp:
+    command: cmd
+  foo:
+    daemon: simple
+  bar:
+    daemon: simple
+    after: [foo]
+`
+
+func (s *serviceControlSuite) SetUpTest(c *C) {
+	s.BaseTest.SetUpTest(c)
+
+	dirs.SetRootDir(c.MkDir())
+
+	s.o = overlord.Mock()
+	s.state = s.o.State()
+
+	s.sysctlArgs = nil
+	systemctlRestorer := systemd.MockSystemctl(func(cmd ...string) (buf []byte, err error) {
+		s.sysctlArgs = append(s.sysctlArgs, cmd)
+		if cmd[0] == "show" {
+			return []byte("ActiveState=inactive\n"), nil
+		}
+		return nil, nil
+	})
+	s.AddCleanup(systemctlRestorer)
+
+	s.serviceMgr = servicestate.Manager(s.state, s.o.TaskRunner())
+	s.o.AddManager(s.serviceMgr)
+	s.o.AddManager(s.o.TaskRunner())
+	s.se = s.o.StateEngine()
+	c.Assert(s.o.StartUp(), IsNil)
+}
+
+func (s *serviceControlSuite) mockTestSnap(c *C) {
+	si := snap.SideInfo{
+		RealName: "test-snap",
+		Revision: snap.R(7),
+	}
+	snaptest.MockSnap(c, servicesSnapYaml1, &si)
+	snapstate.Set(s.state, "test-snap", &snapstate.SnapState{
+		Active:   true,
+		Sequence: []*snap.SideInfo{&si},
+		Current:  snap.R(7),
+		SnapType: "app",
+	})
+
+	err := os.MkdirAll(filepath.Join(dirs.GlobalRootDir, "etc/systemd/system/"), 0775)
+	c.Assert(err, IsNil)
+	err = ioutil.WriteFile(filepath.Join(dirs.GlobalRootDir, "etc/systemd/system/snap.test-snap.foo.service"), nil, 0644)
+	c.Assert(err, IsNil)
+	err = ioutil.WriteFile(filepath.Join(dirs.GlobalRootDir, "etc/systemd/system/snap.test-snap.boo.service"), nil, 0644)
+	c.Assert(err, IsNil)
+}
+
+func (s *serviceControlSuite) TearDownTest(c *C) {
+	dirs.SetRootDir("")
+	s.BaseTest.TearDownTest(c)
```

Comment:
You could also move dirs.SetRootDir to the SetUpTest with:

```go
s.AddCleanup(func() {dirs.SetRootDir("")})
```
and then not need to define TearDownTest at all for serviceControlSuite because the BaseTest.TearDownTest would get called automatically.

I have a slight preference for this, but it's not a big deal if you prefer this style.

Reply:
You're right, I like that, thanks!